src/main/java/io/supertokens/webserver/api/thirdparty/getUsersByEmail/GetUsersByEmailQuery.java

```diff
@@ -0,0 +1,42 @@
+/*
```

Comment:
I don't think this abstraction is required.. no other API is made like this, so I would insist that we stick to existing practices. If you disagree, can you please list down **very clear** benefits of doing something like this, and then if those benefits are indeed much better, then we can (overtime), change all APIs to have this level of abstraction.

Reply:
### Current state
Currently we have 3 layers of application: APIs, Business-logic Functions and Storage Functions.
Imagine a scenario where we decide to add a query parameter to the endpoint. 

All the incoming data is passed as positional argument primitives to functions in Business-logic Functions and Storage Functions, like so:
`(receives a request) -> EmailPassword.findUsers(emails, limit, orderBy, ...) -> StorageLayer.getEmailPassword(...).findUsers(emails, limit, orderBy, ...)`. 

#### Consequence
If we decide to add a parameter, each of these functions has to change signature. So, in more high level terms, change in API causes change of signatures in layers of application that should be isolated from API changes.

### Using Query/Command classes as Argument Objects
Now think about doing such query/command object, which is just an Argument Object (as per uncle bob's Clean Code):
```java
public class FindUsersQuery {
    public List<String> getEmails() {}
    public Optional<Integer> getLimit() {}
    ...
}
```

If we use it, change in API won't cause the change in business logic. We may rename some parameters, introduce new optional ones and it won't cause a cascade of changes in signatures. Hence, we won't have to provide multiple Business-logic Functions and Storage Functions overloads to account for different API variants. This makes the API way clearer.

### Unnecessary abstraction
The fact that we don't yet know what will change in API doesn't deem this an unnecessary abstraction. If inside layers (think Business-logic Functions and Storage Functions) are separated from the API, passing this API input as positional arguments is a weak spot (because it forces multiple layers to change signatures or provide overloads).

### Why I used it
Because I think it's a low-cost, medium-value solution. The parameters for given API call anyway make sense when grouped, not separately.

Reply:
Understood. So this Object would carry all the way through the storage layer right until the params are used in the actual queries?

If not, then I can argue that the benefit this gives is limited to none (cause then only this layer: `(receives a request) -> EmailPassword.findUsers(emails, limit, orderBy, ...)` would be unaffected and StorageLayer functions would have to change anyway)

If yes, then this object would need to be defined in the plugin interface project as the storage layer only depends on the plugin interface and not the core project. Does it make sense to add a new class for each function exposed via the plugin interface?

-------
I assume your final aim is to make is to that if the API's input change, all that needs to change (for most cases) is the input validation in this class and the final SQL query for that API?

Reply:
> So this Object would carry all the way through the storage layer right until the params are used in the actual queries?

That's correct. It prevents from violating open-closed principle all the way down through the application (which happens if signatures need to change). So I would say unless we have a really good reason to do it, I would stick to those query/command objects (regardless of whether validation happens there, main aim is the architectural correctness).

> If yes, then this object would need to be defined in the plugin interface project as the storage layer only depends on the plugin interface and not the core project.

That's true, I noticed it as well. Is there a reason core depends on plugin-interface and not the other way around? To me it would make more sense if plugin-interface classes depended on interfaces defined in core.

> Does it make sense to add a new class for each function exposed via the plugin interface?

This is my first thought. Each function exposed via the plugin interface handles separate usecase. Doing that we can achieve similar level of separation as we have across recipes, which is great over time for managing such APIs.

> I assume your final aim is to make is to that if the API's input change, all that needs to change (for most cases) is the input validation in this class and the final SQL query for that API?

Yes, this way we don't break open-closed principle just to maintain simplicity.

Reply:
> That's true, I noticed it as well. Is there a reason core depends on plugin-interface and not the other way around? To me it would make more sense if plugin-interface classes depended on interfaces defined in core, but I might be missing something.

The plugin-interface is meant for communication between the core and the db plugins. Since it's just an interface, it should not depend on the "thing" that is using the interface or implementing the interface.

----------
One point to note that the plugin interface is not the same as the API interface.. By doing this and adding these new classes which are declared in the plugin-interface and created in the API classes is a bit weird.. If the API changes, then we wouldn't want the plugin-interface class to change (unless necessary).